crates/core/src/db/datastore/locking_tx_datastore/datastore.rs

```diff
@@ -417,45 +417,69 @@ impl MutTxDatastore for Locking {
     }
 }
 
-pub(crate) fn record_metrics(ctx: &ExecutionContext, tx_timer: Instant, lock_wait_time: Duration, committed: bool) {
-    let workload = &ctx.workload();
-    let db = &ctx.database();
-    let reducer = ctx.reducer_name();
-    let elapsed_time = tx_timer.elapsed();
-    let cpu_time = elapsed_time - lock_wait_time;
-
-    let elapsed_time = elapsed_time.as_secs_f64();
-    let cpu_time = cpu_time.as_secs_f64();
-    // Note, we record empty transactions in our metrics.
-    // That is, transactions that don't write any rows to the commit log.
-    DB_METRICS
-        .rdb_num_txns
-        .with_label_values(workload, db, reducer, &committed)
-        .inc();
-    DB_METRICS
-        .rdb_txn_cpu_time_sec
-        .with_label_values(workload, db, reducer)
-        .observe(cpu_time);
-    DB_METRICS
-        .rdb_txn_elapsed_time_sec
-        .with_label_values(workload, db, reducer)
-        .observe(elapsed_time);
-
-    let mut guard = MAX_TX_CPU_TIME.lock().unwrap();
-    let max_cpu_time = *guard
-        .entry((*db, *workload, reducer.to_owned()))
-        .and_modify(|max| {
-            if cpu_time > *max {
-                *max = cpu_time;
-            }
-        })
-        .or_insert_with(|| cpu_time);
-
-    drop(guard);
-    DB_METRICS
-        .rdb_txn_cpu_time_sec_max
-        .with_label_values(workload, db, reducer)
-        .set(max_cpu_time);
+/// Holds sufficient information to record transaction metrics on drop.
+pub struct RecordTxMetricsOnDrop<'a> {
```

Comment:
This is very similar to what I am trying to do with iter metrics i.e to batch and delegate to later. I will rebase on your branch to make it consistent across code.